### ISSUE TITLE
REST API(POST/PUT/GET/DELETE) help required

### ISSUE THREAD

**Issue by sureshkvl** [corenova/yang-forge#31]
Hi Peter,

I am implementing the sample students yang model, and want to try POST/PUT/GET/DELETE REST API endpoints.

https://github.com/sureshkvl/yangforge-examples/blob/master/students.yang

I am able to run the students.yang as below,
yfc run students.yang
1. POST API.
   I am able to POST the student in the below API
   POST - http://localhost:5000/restjson/students-yang/students-container/students-list
2. GET API
   I am able to get the students lists as below,
   http://localhost:5000/restjson/students-yang/students-container/students-list

output :

[
  {
    "name": "ramesh",
    "rollno": 22,
    "address": "bangalore,india"
  },
  {
    "name": "suresh",
    "rollno": 34,
    "address": "chennai,india"
  }
]

Query:   How to get the specific index or key student object from the student list.?
Basically i want to get only "suresh" student details from the list.
1. PUT API:
   What is the API(URL) to modify the specific student  details?
   http://localhost:5000/restjson/students-yang/students-container/students-list/xxxxxxx
2. DELETE API:
   Is it possible to delete the object from the list..?

Thanks
suresh

**Comment by sekur** [comment]
Hi Suresh,

On the students.yang schema, the "key" for the students-list should be "rollno" instead of "rollnumber". 

Then you can use the value of "rollno" as the key to get the specific student entry and operate on it:

GET/PUT/DELETE http://localhost:5000/restjson/students-yang/students-container/students-list/22

In terms of convention, it would be better to have:

container school {
  list students {
    ...
  }
}

Peter Lee

**Comment by sureshkvl** [comment]
Hi Peter,
Thanks. I have updated the yang model
https://github.com/sureshkvl/yangforge-examples/blob/master/students.yang

1) Now Index based GET API is working.
GET  http://localhost:5000/restjson/school-yang/school/students/31

{
  "name": "ramesh kumar",
  "rollno": 31,
  "address": "chennai,india"
}

2) DELETE REST API throws the below error.
 DELETE   http://localhost:5000/restjson/school-yang/school/students/31
{
  "error": "TypeError: Object  has no method 'then'"
}
1. PUT API   http://localhost:5000/restjson/school-yang/school/students/31
   returns the below error (404 Not Found)
   Cannot PUT /restjson/school-yang/school/students/31

Thanks
suresh

**Comment by sekur** [comment]
Take a look at ListRouter inside features/restjson.coffee

See if you can identify the issue(s).

I will fix it tomorrow and you can reference the commit to confirm.

**Comment by sekur** [comment]
@sureshkvl - please pull latest from GitHub master and give it a try.

**Comment by sureshkvl** [comment]
Hi Peter,

I have tested with the latest commit.
1. POST is working.
2. GET the complete List is working  (GET http://localhost:5000/restjson/school-yang/school/students)
3. Index based GET API is NOT working.(GET http://localhost:5000/restjson/school-yang/school/students/33)
   Returns Error : 404
   Cannot GET /restjson/school-yang/school/students/33
4. PUT API is also NOT working.  Similar to GET API error
5.  DELETE API returns the below output (it is not deleting the record, instead of listing the complete object)
   {
   "school": {
     "students": [
       {
         "name": "suresh123",
         "rollno": 33,
         "address": "chennai,india"
       }
     ]
   }
   }

Thanks
suresh

**Comment by sekur** [comment]
Ok, just tested with your schema and can reproduce the issue.

The problem is because the "key" is coming in as a "string" via restjson request but internally it is stored as uint32. 

I will need to make a fix on data-synth library to do key value format conversion during list entry lookup.

**Comment by sekur** [comment]
Ok, the latest commit updated to data-synth 0.7.14.  After pull from master, should do npm install to grab latest data-synth. 

The fix will resolve GET and PUT by key (that is numeric).  I'll need to take another look to handle DELETE.

**Comment by sureshkvl** [comment]
Hi Peter,
I just tested with updated data-synth.
Yes. GET & PUT by key works fine.

Thanks
suresh
